### ISSUE TITLE
Bug: [XOR] Type checking broken in v10 

### ISSUE THREAD

**Issue by papermana** [ts-essentials/ts-essentials#400]
### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `typescript`                           | `5.4.5` |
| `ts-essentials`                        | `10.0.0` |

### Repro Code

```typescript
It's available in the Typescript Playground I linked to.
```

### TypeScript Playground

https://www.typescriptlang.org/play/?#code/PTAEGEHsAcEsFMAmoBmAnSBbUAXAzgLTx57wB2OsAhgDZ6gBu8aARlZZgFA4Ce08oAOqwcAC0gBXHAB4AKn3gBGADSh5-AEwA+UAF5QAb06hQAbQAKoWGVABRAB4BjGhMTxpAa3g9IKNQpVQLx8-dXhtAF0AfgAuUDJ4JjQAbk4AX1TOEFAADQB5ACVcUXZQKkdHeGh8YrR4AV5+UGgqNCpMeBxmeioyZEc69mJQROYeZsgSWBYaAUcsFmt2WEgbX2L4WDRmjH40SmIYrLACHZhmA-p0LFwFUECqegB3eBoaMp6ycbJVgnmyFA0WCOShkADmZz2l1QGGwjWGGjKfVAAGZjqBTtBdhcEFdYbcmhpVI9QC83h8kd9fv9AcDQRCsed9riYTd4fRFEjkGjspjsczhtc4XcUcTnq93iTelSyH9VrSQdYGfzoUKCcNOb1kBp0QAxSDbMQCPBULygdbOVoiHjE8liOoNO4tNodLpoeglJi4SCgFgCRkMWBuRCqUY2WB+VY0cY4J4+1oCBKVEitHgAOm4d3yBRRcgCqjCRP8-BROn0AAphGJJDJC6AAGTF+CipuKHSNsKKACUoAAPqBKyJxFI8-xOR2FC3C+2mxoe-3B9WR52G7OC5OZ2EUV3Mtls8VSuVKtV6Pb6urmq12p1ur6pOayNHQGCJFeKPVT6J-XVSGRKua-EZKEWTVWMfXWI1SHVPAjl5SEcUFfF4XuMVSQlClpXial5SBRVwXggU8TZBR6ERLVUXRPkmVVJC7iLEkyUlT4ZTlAFcPpAiaOI-gOS5CjsnMFUQNopoUVATAqHGP0dmIcgcFUFh7yNOoyhUn5n1fNp32GCS3DKFgaw2TAM2yWRRFgegLKwnBQDwLBOnM-CxFKOMJBoZBpLcPAtioGYBGsUBBiBXgFPvERbOYJgek+EZ7HaaBZgzZDsw0UclHXTQMubMsByrYdawUREJxLLK21XTt51yocazS8cmynQrN0KndOHRcxHlIU87kcL9HA8JUFPgRwqAkKCUo0t8umGZybM2ZSynJICEPodYAANO1W81tlod4lsIgDQHWwrNqeWByWkrFZIoDN-jwGyujuxEKwMFBIEgBTWjSOIUukF63riO60CVNJVAMNg0ABnAgfBEHQAkMgPB+J4yC0LtIehiFdB0IwTAjAd4XWV6fV0EnQAAckBpUyZ7OocAkNAbCJ1ITFp+mbHB1I0la7IABFIGGH4bJaEgLx6ob+vBQbhtGgRszEl9Jo-A8bJ2zjhJuVYBAgr8oPZUkzveC6fzkuIjrHVbVDN8ILcO+HEcgZHNvhEywAAZWsf9VrtpGyE24abE9ARMLwEp+FUI1Rd6rxtn9wWJq0qblcpTjxlO8lRm2I2rpsthSGQAArUabNWDYZN-f91kgea9suG7Vju3BiBwMTnqJj60C+3JClzP7IHR4HQfB-uYdUb2HZRtHbKhpU9Gx4wrD8csCb8Im9FJinp-BanAs6NnUDe5md7phnfVaTmgA

### Expected Result

I expected `XOR<Foo, Bar>` to allow narrowing down the type to `Foo` if I can show that `Bar` doesn't apply.

### Actual Result

The type of `XOR<Foo, Bar>` cannot be narrowed down to `Foo` by checking for properties from `Bar`, because there's up to 50 other `unknown` types in the union.

**Comment by Beraliv** [comment]
Hey @papermana. Thank you for reporting the issue and your patience.

After adding `Prettify` (i.e. https://tsplay.dev/WKZdzm), I see a regression: `{
    bar?: undefined;
    foo?: undefined;
}` shouldn't be a part of any XOR output

I will add the fix and will cover it with the unit test so it's not repeated next time

I will ping you when it's ready, should be done in ~3-4 hours

**Comment by Beraliv** [comment]
@papermana This would be the fix

```ts
type IsAny<Type> = 0 extends 1 & Type ? true : false;

type IsUnknown<Type> = IsAny<Type> extends true ? false : unknown extends Type ? true : false;

type Without<Type1, Type2> = IsUnknown<Type2> extends true
  ? never
  : { [P in Exclude<keyof Type1, keyof Type2>]?: never };
```

Playground with a fixed version - https://tsplay.dev/w19z8w

**Comment by Beraliv** [comment]
Before merging the fix https://github.com/ts-essentials/ts-essentials/pull/401, kindly ask to verify if it fixes the problem in your project. If there are no objections and changes look good, I will release a patch version 10.0.1 tomorrow as a part of https://github.com/ts-essentials/ts-essentials/pull/398

**Comment by papermana** [comment]
@Beraliv I can confirm this fixes the issue in our codebase. Thank you for such an incredibly fast response!

**Comment by Beraliv** [comment]
I have tried to cut a release but there is some issue with the release GHA, I will have a look at it later today.

**Comment by Beraliv** [comment]
@papermana the fix has been released as [v10.0.1](https://github.com/ts-essentials/ts-essentials/releases/tag/v10.0.1), therefore closing the issue. Please let me know if you need any help!
